python/cudf_polars/cudf_polars/dsl/expr.py

```diff
@@ -0,0 +1,1023 @@
+        Collect information about aggregations in groupbys.
+
+        Parameters
+        ----------
+        depth
+            The depth of aggregating (reduction or sampling)
+            expressions we are currently at.
+
+        Returns
+        -------
+        Aggregation info describing the expression to aggregate in the
+        groupby.
+
+        Raises
+        ------
+        NotImplementedError if we can't currently perform the
+        aggregation request (for example nested aggregations like
+        ``a.max().min()``).
+        """
+        raise NotImplementedError
+
+
+def with_mapping(fn):
+    """Decorate a callback that takes an expression mapping to use it."""
+
+    def _(
+        self,
+        df: DataFrame,
+        *,
+        context=ExecutionContext.FRAME,
+        mapping: dict[Expr, Column] | None = None,
+    ) -> Column:
+        """Look up self in the mapping before evaluating it."""
+        if mapping is None:
+            return fn(self, df, context=context, mapping=mapping)
+        else:
+            try:
+                return mapping[self]
+            except KeyError:
+                return fn(self, df, context=context, mapping=mapping)
+
+    return _
+
+
+class NamedExpr(Expr):
+    __slots__ = ("name", "children")
+    _non_child = ("dtype", "name")
+
+    def __init__(self, dtype: plc.DataType, name: str, value: Expr) -> None:
+        super().__init__(dtype)
+        self.name = name
+        self.children = (value,)
+
+    @with_mapping
+    def evaluate(
+        self,
+        df: DataFrame,
+        *,
+        context: ExecutionContext = ExecutionContext.FRAME,
+        mapping: dict[Expr, Column] | None = None,
+    ) -> Column:
+        """Evaluate this expression given a dataframe for context."""
+        (child,) = self.children
+        return Column(
+            child.evaluate(df, context=context, mapping=mapping).obj, self.name
+        )
+
+    def collect_agg(self, *, depth: int) -> AggInfo:
+        """Collect information about aggregations in groupbys."""
+        (value,) = self.children
+        return value.collect_agg(depth=depth)
+
+
+class Literal(Expr):
+    __slots__ = ("value",)
+    _non_child = ("dtype", "value")
+    value: pa.Scalar
+
+    def __init__(self, dtype: plc.DataType, value: Any) -> None:
+        super().__init__(dtype)
+        self.value = pa.scalar(value)
+
+    @with_mapping
+    def evaluate(
+        self,
+        df: DataFrame,
+        *,
+        context: ExecutionContext = ExecutionContext.FRAME,
+        mapping: dict[Expr, Column] | None = None,
+    ) -> Column:
+        """Evaluate this expression given a dataframe for context."""
+        # TODO: obey dtype
+        obj = plc.interop.from_arrow(self.value)
+        return Scalar(obj)  # type: ignore
+
+    def collect_agg(self, *, depth: int) -> AggInfo:
+        """Collect information about aggregations in groupbys."""
+        raise NotImplementedError("Literal in groupby")
```

Comment:
These seems redundant, same as the top-level behavior. Is this meant to be a NotImplementedError, or is there a different exception we should raising here to indicate that this will never be supported? Or is this just providing a slightly more informative exception, in which case we could probably just change the definition in the parent class's definition to be something like `raise NotImplementedError(f"{self.__class__.__name__} in groupby")`.

Reply:
I have a followup where I implement this.